### ISSUE TITLE
Dropping the main branch

### ISSUE THREAD

**Issue by maehler** [genomic-medicine-sweden/poppy#56]
## TL;DR

I want to:

- Trash the main branch
- Rename dev &rarr; main
- Have release-please act on new main
- Encourage squash-merges of PRs

This would help with:

- Keeping a cleaner history
- Reduce the amount of duplicate work

For more context, keep reading. Discussion encouraged.

## The issue

Here is something that has been bothering me for a while. In the current setup of the repo we have two long-running branches: dev and main. The idea is that changes from dev are merged into main and release-please acts on this branch to generate new release PRs.

My main issue with this is that it creates a lot of extra work, basically for nothing. When dev should be merged into main, someone has to open an PR, and someone else has to review it. The thing is that the changes that are being merged have already been reviewed before being merged into dev, so this is a bunch of work that essentially already has been done. Given the branch protection rules, no other commits would happen on dev&mdash;all would have to pass through PRs and associated code reviews.

Another issue I have with this is that it creates a messy history. For release-please to work as intended, each commit should have a specific format. Merging from dev to main for new releases means that we cannot do squash-merges if we have more than one change that should be reflected in the changelog. A merge commit would have to be created, and these are just commits without real meaning and more or less just represents noise in this context.

Furthermore, if the release PR introduces any changes to the codebase, these are not readily reflected in the dev branch once the PR has been merged. For this to happen, main has to be merged into dev. As a consequence, that also means that all the merge commits that have been done on the main branch will also be pulled into the dev branch, making things even noisier since they are not at all relevant in this context.

## My proposed approach

My suggestion is to remove the current main branch and rename the dev branch to main. Current branch protection rules would still apply, so new features and bug fixes would still have to be merged via PRs. The main difference is that these changes wouldn't have to be merged again, but releases could be made directly from the new main branch whenever we see fit. As long as there are well-formatted commits since the last release, a release PR will be open, and creating a new release is done simply by merging this PR.

My preferred approach with PRs is to squash-merge them once they have been approved, at least if the PR relates to a relatively well-defined issue or featurs&mdash;something that can be described in a single commit message (50 chars). The main advantage of this is that we get a cleaner history. Commits on a feature branch that are related to trial and error will never be seen on the main branch, and why would they?

For example, let's say that when working on a feature, you accidentally introduce a bug. On the same branch you fix that bug with the commit message "fix: bug related to x". If we would use a merge commit, this commit would end up on the main branch and also the changelog once a new release is created. The thing is that this is a bug that never would manifest itself on the main branch, since it was specific to an issue on the feature branch. By doing a squash-merge instead, all of this is baked into a single commit and treated as a single unit. This will require a bit more thought when creating and working on PRs, but this is probably not a bad thing. I think PRs should only live as long as it's necessary for a feature to be implemented or a bug to be fixed, and any commits on that branch should only relate to the feature or bug in question.

I realize that this can sometimes become a bit tedious. For example, let's say you're working on a feature and see that things break due to something outside the scope of the feature in question. If this is a simple fix (e.g. bumping some dependency or fixing a spelling error or a conditional that ended up backwards), then it is tempting to not open a new PR just for that and just fix it in-place. In these cases, I think it's ok to fix this, but then it is also important to realize that this is a fix that is related to the state of the main branch as well, and should be reflected there too. If we squash merge, the commit fixing this issue will disappear, but it release-please makes it possible to still include this change in the release notes/changelog by [describing the fix in the commit message of the squashed commit](https://github.com/googleapis/release-please?tab=readme-ov-file#what-if-my-pr-contains-multiple-fixes-or-features). An example of this can be seen in https://github.com/hydra-genetics/cnv_sv/commit/1c960c638ffdefea8b23296f8a49fbe65b82b557 where I was working on a feature, and stumbled upon the pulp issue that we've been seeing recently. I fixed that in the same PR, and in the commit message I added "fix: pin pulp version to <2.8.0", which then would end up in the changelog for the next release.

Then there are also cases when things get out of hand in a PR and you for some reason end up with a bunch of commits that don't necessarily make sense to bunch together, but at the same time the work of splitting these off into their own branches would be messy and time-consuming. In cases like these I can live with the PR being merged with a merge commit instead. This happened to me the other day where I had a [bunch of commits](https://github.com/genomic-medicine-sweden/poppy/pull/52) that were all related to Github Actions in some way, but different aspects of it that I didn't want to bunch together. Here I swallowed my pride and didn't squash-merge, but this also assumes that all the commit messages in the PR are formatted correctly. If they are not, it doesn't really matter. Then it's better to either do a squash-merge and describe the changes made in the commit message (as [described in the release-please documentation](https://github.com/googleapis/release-please?tab=readme-ov-file#what-if-my-pr-contains-multiple-fixes-or-features)), or go back and do an interactive rebase to update the commit messages and then force push to the branch in question before merging the PR.

These were some thoughts, and I would appreciate any feedback on this. I realize it became somewhat of a rant, but this has been bugging me for quite some time now. The current approach is what is used in many (all?) repos related to hydra-genetics, and I thought that this repo would suffer the least amount of disruption by applying these changes since there are no releases yet, and the main branch is practically empty. Let me know what you think.

Thanks for coming to my TED talk.

**Comment by elleira** [comment]
I think it sounds good (despite absolutely not being what my next pull request will look like (due to complete scatter-brain at the moment :sweat_smile: )! But then we should have two approves to be able to merge directly in main then?

**Comment by maehler** [comment]
Regarding your PR, no worries at all! I know the feeling, and sometimes things just have to evolve organically to get the stuff out there.

I don't think that two approvals would be needed for the merging, necessarily. At the moment, the dev branch is the default, so whenever anyone pulls the repo they will still get stuff that only had one approval. I think these merges can stay at one approval, but perhaps it could be an idea to require the release PRs to have two approvals instead? This could help make sure that we agree that whatever will be included in the release is solid.

**Comment by maehler** [comment]
`main` is dead, long live `main`.

To update any local copies:

```bash
git branch -m dev main
git fetch origin
git branch -u origin/main main
git remote set-head origin -a
```

Once the PR for fixing Github Actions is merged (https://github.com/genomic-medicine-sweden/poppy/pull/60), this issue will be closed.
